File: Algorithm/kosaraju.py
from collections import defaultdict


class Graph:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjacency_list = defaultdict(list)
        self.reverse_adjacency_list = defaultdict(list)

    def add_edge(self, u, v):
        self.adjacency_list[u].append(v)
        self.reverse_adjacency_list[v].append(u)

    def kosaraju(self):
        stack = []
        visited = [False] * self.vertices

        def fill_order(u):
            visited[u] = True
            for neighbor in self.adjacency_list[u]:
                if not visited[neighbor]:
                    fill_order(neighbor)
            stack.append(u)

        for i in range(self.vertices):
            if not visited[i]:
                fill_order(i)

        visited = [False] * self.vertices
        print("Strongly Connected Components:")

        def dfs_reverse(u):
            visited[u] = True
            print(u, end=' ')
            for neighbor in self.reverse_adjacency_list[u]:
                if not visited[neighbor]:
                    dfs_reverse(neighbor)

        while stack:
            u = stack.pop()
            if not visited[u]:
                print("SCC: ", end='')
                dfs_reverse(u)
                print()


g = Graph(8)
g.add_edge(0, 1)
g.add_edge(1, 2)
g.add_edge(2, 3)
g.add_edge(2, 4)
g.add_edge(3, 0)
g.add_edge(4, 5)
g.add_edge(5, 6)
g.add_edge(6, 4)
g.add_edge(6, 7)

print("Graph edges: 0->1, 1->2, 2->3, 2->4, 3->0, 4->5, 5->6, 6->4, 6->7")
g.kosaraju()

File: Algorithm/kosaraju.cpp
#include <iostream>
#include <vector>
#include <stack>

class Graph {
public:
    int vertices;
    std::vector<std::vector<int>> adjacencyList;
    std::vector<std::vector<int>> reverseAdjacencyList;

    Graph(int v) : vertices(v), adjacencyList(v), reverseAdjacencyList(v) {}

    void addEdge(int u, int v) {
        adjacencyList[u].push_back(v);
        reverseAdjacencyList[v].push_back(u);
    }

    void kosaraju() {
        std::stack<int> st;
        std::vector<bool> visited(vertices, false);

        for (int i = 0; i < vertices; ++i) {
            if (!visited[i]) {
                fillOrder(i, visited, st);
            }
        }

        visited.assign(vertices, false);
        std::cout << "Strongly Connected Components:" << std::endl;

        while (!st.empty()) {
            int u = st.top();
            st.pop();

            if (!visited[u]) {
                std::cout << "SCC: ";
                dfsReverse(u, visited);
                std::cout << std::endl;
            }
        }
    }

private:
    void fillOrder(int u, std::vector<bool>& visited, std::stack<int>& st) {
        visited[u] = true;

        for (int neighbor : adjacencyList[u]) {
            if (!visited[neighbor]) {
                fillOrder(neighbor, visited, st);
            }
        }

        st.push(u);
    }

    void dfsReverse(int u, std::vector<bool>& visited) {
        visited[u] = true;
        std::cout << u << " ";

        for (int neighbor : reverseAdjacencyList[u]) {
            if (!visited[neighbor]) {
                dfsReverse(neighbor, visited);
            }
        }
    }
};

int main() {
    Graph g(8);
    g.addEdge(0, 1);
    g.addEdge(1, 2);
    g.addEdge(2, 3);
    g.addEdge(2, 4);
    g.addEdge(3, 0);
    g.addEdge(4, 5);
    g.addEdge(5, 6);
    g.addEdge(6, 4);
    g.addEdge(6, 7);

    std::cout << "Graph edges: 0->1, 1->2, 2->3, 2->4, 3->0, 4->5, 5->6, 6->4, 6->7" << std::endl;
    g.kosaraju();

    return 0;
}
